src/indicators/drought/wrsi.py

```diff
@@ -0,0 +1,309 @@
+from typing import List, Literal, Union, get_args
+from urllib.error import HTTPError
+from urllib.request import urlopen
+from zipfile import ZipFile
+
+import numpy as np
+import pandas as pd
+import rioxarray  # noqa: F401
+import xarray as xr
+from rasterio.crs import CRS
+
+logger = logging.getLogger(__name__)
+
+_BASE_URL = (
+    "https://edcftp.cr.usgs.gov/project/fews/"
+    "africa/west/dekadal/wrsi-chirps-etos/{region}/"
+)
+
+_URL_ZIP_FILENAME = "w{year:04}{dekad}{region}.zip"
+
+_BASE_FILENAME = "w{year:04}{dekad}{type}.tif"
+
+
+def _get_url_path(year, dekad, region):
+    if region == "cropland":
+        url_region, file_region = "west", "wa"
+    else:
+        url_region, file_region = "west1", "w1"
+
+    url_zip_filename = _URL_ZIP_FILENAME.format(
+        year=year, dekad=dekad, region=file_region
+    )
+
+    base_url = _BASE_URL.format(region=url_region)
+    return base_url + url_zip_filename
+
+
+def _download_wrsi(year: int, dekad: int, region: str, raw_dir: Path):
+    try:
+        url = _get_url_path(year=year, dekad=dekad, region=region)
+        resp = urlopen(url)
+    except HTTPError:
+        logger.error(
+            f"No data available for region {region} for "
+            f"dekad {dekad} of {year}, skipping."
+        )
+        return
+
+    zf = ZipFile(BytesIO(resp.read()))
+    for type in ["do", "eo", "er"]:
+        file_name = _BASE_FILENAME.format(year=year, dekad=dekad, type=type)
+        save_path = Path(raw_dir, file_name)
+        save_path.unlink(missing_ok=True)
+
+        zf.extract(file_name, raw_dir)
+
+
+_RAW_PATH = Path(
+    os.getenv("AA_DATA_DIR"), "public", "raw", "glb", "wrsi", "west_africa"
+)
+
+
+def _get_raw_dir(region):
+    return Path(_RAW_PATH, region)
+
+
+def _date_to_dekad(date_obj: Union[date, str]):
+    if isinstance(date_obj, str):
+        date_obj = date.fromisoformat(date_obj)
+    year = date_obj.year
+    dekad = (
+        (date_obj.timetuple().tm_mday // 10) + ((date_obj.month - 1) * 3) + 1
+    )
+    return year, dekad
+
+
+def _dekad_to_date(year: int, dekad: int):
+    month = ((dekad - 1) // 3) + 1
+    day = 10 * ((dekad - 1) % 3) + 1
+    return datetime.datetime(year, month, day)
+
+
+def _fp_date(fp):
+    return _dekad_to_date(int(fp[1:5]), int(fp[5:7]))
+
+
+RegionArgument = Literal["cropland", "rangeland"]
+_VALID_REGION = get_args(RegionArgument)
+
+
+def download_wrsi(
+    region: RegionArgument,
+    start_date: Union[date, str, List[int], None] = None,
+    end_date: Union[date, str, List[int], None] = None,
+    redownload: bool = False,
+):
+    if region not in _VALID_REGION:
+        raise ValueError("`region` must be one of 'cropland' or 'rangeland'.")
+
+    if start_date is None:
+        start_year, start_dekad = 2001, 13
+    elif isinstance(start_date, list):
+        start_year, start_dekad = start_date[0], start_date[1]
+    else:
+        start_year, start_dekad = _date_to_dekad(start_date)
+
+    if isinstance(end_date, list):
```

Comment:
The `end_date` structure is slightly amended just because if it isn't a `list`, then both methods need to call `_date_to_dekad()`, so wanted to not repeat myslf by calling it twice. But maybe still clearer to just match and do some slight repetition? What do you think?

Reply:
Aah I see, missed that. Then yes this is fine. Could maybe with the start_date to first `if list` and then `elif is None` to mimic the structure a bit more, but that is very minor.